newIDE/app/src/EventsBasedBehaviorEditor/EventsBasedBehaviorEditorDialog.js

```diff
@@ -26,9 +28,14 @@ export default class EventsBasedBehaviorEditorDialog extends React.Component<
       project,
     } = this.props;
 
+    const onApplyAndFixIssues = () => {
+      onApply();
+      fillRequiredBehaviorProperties(this.props.project);
+    };
```

Comment:
> What is updating the property list of behaviors in objects to follow the changes done on the extension?

I would say: nothing.

If you add a new property, it's not added in the "behavior content" (aka: behavior data) of objects having this behavior.
This is not a problem for events or the generated behavior because a default value is used if the behavior data is not containing the key:
https://github.com/4ian/GDevelop/blob/287ebd4244ecb9ba272b648dcf75a7e029d190ed/GDJS/GDJS/Events/CodeGeneration/BehaviorCodeGenerator.cpp#L169-L175

How is this affecting you? It means that when the dialog close, you must fill the required behavior properties in the whole project, but that should not be a problem?

Reply:
> How is this affecting you? It means that when the dialog close, you must fill the required behavior properties in the whole project, but that should not be a problem?

`WholeProjectRefactorer::FindInvalidRequiredBehaviorProperties` relies on the `behaviorContent` to find required behaviors that are not filled correctly, but these property keys doesn't exist yet on objects so the issues fly under the radar. I could use `propertyDescriptors` to check this case but it feels almost like an encapsulation breach to do so. 

> > What is updating the property list of behaviors in objects to follow the changes done on the extension?
>
> I would say: nothing.

What is really strange is that the issues are spotted the next time I close the behavior settings dialog. Something must have the side effect of refreshing the properties. Otherwise, I guess they would not appear on the object editor either.

Reply:
I see, this might be because you call GetProperties to get all the properties (and you should use the property descriptor to get the type of the property, to check if it's "behavior"), but this is only ready after regenerating the extension... Which will be done later. I'll think about that.

Reply:
I think this is what you're searching for the "thing that does side effect when properties are changed": https://github.com/4ian/GDevelop/blob/916938b4a8df88a060da95019b4921d615f15205/newIDE/app/src/MainFrame/EditorContainers/EventsFunctionsExtensionEditorContainer.js#L63

Not sure yet what's the best way to ensure this is run before refactoring (and if we should even wait for this or not)

Reply:
Ok, I attempted to force a refresh on _editBehavior to see if it was it, but I probably didn't do the right thing. I will retry.

EventsBasedBehaviorEditorDialog:
```
    const onApplyAndFixIssues = () => {
      // TODO It only works when the dialog is opened and closed again.
      onApply();
      if (fillRequiredBehaviorProperties(this.props.project)) {
        onApply();
      }
    };

    return (
      <Dialog
        onApply={onApplyAndFixIssues}
```

EventsFunctionsExtensionEditor/index:
```
          <EventsBasedBehaviorEditorDialog
            project={project}
            eventsFunctionsExtension={eventsFunctionsExtension}
            eventsBasedBehavior={editedEventsBasedBehavior}
            onApply={() => {
              if (this.props.unsavedChanges)
                this.props.unsavedChanges.triggerUnsavedChanges();
              this._editBehavior(null);
            }}
```

```
  _editBehavior = (editedEventsBasedBehavior: ?gdEventsBasedBehavior) => {
    this.setState(
      state => {
        // If we're closing the properties of a behavior, ensure parameters
        // are up-to-date in all event functions of the behavior (the object
        // type might have changed).
        if (state.editedEventsBasedBehavior && !editedEventsBasedBehavior) {
          gd.WholeProjectRefactorer.ensureBehaviorEventsFunctionsProperParameters(
            this.props.eventsFunctionsExtension,
            state.editedEventsBasedBehavior
          );
        }

        return {
          editedEventsBasedBehavior,
        };
      },
      () => {
        if (!editedEventsBasedBehavior) {
          // If we're closing the properties of a behavior, notify parent
          // that a behavior was edited (to trigger reload of extensions)
          if (this.props.onBehaviorEdited) {
            this.props.onBehaviorEdited();
          }

          // Reload the selected events function, if any, as the behavior was
          // changed so objects containers need to be re-created. Notably, the
          // type of the object that is handled by the behavior may have changed.
          if (this.state.selectedEventsFunction) {
            this._loadEventsFunctionFrom(
              this.props.project,
              this.state.selectedEventsFunction
            );
          }
        }
      }
    );
```

EventsFunctionsExtensionEditorContainer:
```
        <EventsFunctionsExtensionEditor
          key={eventsFunctionsExtension.ptr}
          project={project}
          eventsFunctionsExtension={eventsFunctionsExtension}
          setToolbar={this.props.setToolbar}
          resourceSources={this.props.resourceSources}
          onChooseResource={this.props.onChooseResource}
          resourceExternalEditors={this.props.resourceExternalEditors}
          openInstructionOrExpression={this.props.openInstructionOrExpression}
          onCreateEventsFunction={this.props.onCreateEventsFunction}
          initiallyFocusedFunctionName={initiallyFocusedFunctionName}
          initiallyFocusedBehaviorName={initiallyFocusedBehaviorName}
          onBehaviorEdited={this._onBehaviorEdited}
```

Reply:
`loadProjectEventsFunctionsExtensions` from `EventsFunctionsExtensionsLoader` is called asynchonously and happen after `fillRequiredBehaviorProperties`. That explains why it only woks the 2nd time the dialog is closed, but it seems to need a big change to run `fillRequiredBehaviorProperties` after it.

Reply:
> That explains why it only works the 2nd time the dialog is closed, but it seems to need a big change to run fillRequiredBehaviorProperties after it.

Correct that's a pretty big change. Extensions are reloaded/code generated "later", when needed in theory.

> But maybe it would be easier and cleaner to make the property keys always consistent with the behavior definition?

Can you expand on what you mean exactly by that ("property keys")?

Reply:
I wonder if, still, the best solution is to force the regeneration (re-declaration precisely) of extensions and wait synchronously for it to be done (with a Promise) before doing the refactoring of required properties... 🤔

Reply:
> Can you expand on what you mean exactly by that ("property keys")? I guess you mean `behavior.GetProperties`?

Yes, what I meant is that `GetProperties` uses the `behaviorContent.GetContent()` to build the key/value list, but it could use `eventsBasedBehavior.getPropertyDescriptors()` or an equivalent from the base class to build the keys. This would avoid inconsistencies.

Reply:
I believe we should run the refactoring with the proper declared extension and behavior metadata - no point trying to do something before that because the metadata would not be correct and this could create inconsistencies in the future. I *think* therefore we can do something like this:

- Change `onBehaviorEdited` so that it returns a promise
- In `_onBehaviorEdited`, change it so that it returns the result of `onLoadEventsFunctionsExtensions`
- In `MainFrame`, change `onLoadEventsFunctionsExtensions` so that it returns the result of `loadProjectEventsFunctionsExtensions` - which is a promise that resolves when extensions are reloaded and code generated. 
  - Note that we don't care about code generation, even if we'll have to wait for it (could be improved in the future)
- move `fillRequiredBehaviorProperties` to `EventsFunctionsExtensionEditor/index.js` (because I believe this is something fairly high level that should not be buried too much).
- Call it when `onBehaviorEdited` resolves or reject (it can only reject if extensions code generation failed, like on the web-app, but **even** in this case the metadata of extensions are valid and should be used for the refactoring).

With this, we should run the refactoring as soon as extensions are ready.
In the future, check if it works (and if this does not create too much lag?). In the future would be great to show an indicator to show that these things are done in the background (and to really do that in a background "worker") (but that's for another day).

If this works, I believe we'll be in a good state for this?